### PR TITLE
Docker provider

### PR DESCRIPTION
**Pull request by errygg**
Reorganized the examples after adding a docker_provider example. Updated the README.md files as well to segregate the different providers.

### CODE REVIEW

**Comment by ncs-alane** [comment]
@errygg can you point this pull request to `ncs-alane-0.4.0`?

File: examples/docker_provider/Gemfile [hunk 1]
```diff
@@ -0,0 +1,5 @@
+source 'http://rubygems.org/'
+
+gem 'test-kitchen'
+gem 'kitchen-terraform'
```
Comment (ncs-alane) [inline examples/docker_provider/Gemfile]: If we change this to source from the root of the project directory then we can drop dependency on `test-kitchen` and this will example can be used as additional integration tests.

File: examples/docker_provider/Gemfile [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+gem 'kitchen-dokken'
```
Comment (ncs-alane) [inline examples/docker_provider/Gemfile]: We could pull `kitchen-dokken` in to the project and create a transport plugin that can handle SSH and Docker. ^_^

Comment (errygg) [inline examples/docker_provider/Gemfile]: That was my intent, forgot to remove kitchen-dokken from the Gemfile. I think we should probably remove that dependency for now and open a new feature for adding kitchen-dokken capability.

File: examples/docker_provider/README.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+The provisioner defines a variable file, [testing.tfvars], that includes the Docker configuration variables.
+[testing.tfvars]: testing.tfvars
+
+### Platforms
+This section is irrelevant, but must be populated.
+
+### Transport
+The transport defines the mechanism for interacting with the Docker container.
+
+ssh with password authentication is the mechanism defined here. The Docker container that will be used for this example runs the ssh daemon in the foreground in order to allow the verifier to execute tests on the running container.
```
Comment (ncs-alane) [inline examples/docker_provider/README.md]: "ssh" => "SSH"

File: examples/docker_provider/README.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+### Verifier
+The verifier defines a single `default` group that includes a single test for the `operating_system` on the `localhost` using port number `2222`. This port is mapped to the docker container's ssh port `22`.
+
+### Suites
+The suite name corresponds to the integration test directory pathname as is typical for other test kitchen plugins.
```
Comment (ncs-alane) [inline examples/docker_provider/README.md]: Other plugin behaviour seems irrelevant to the example.

File: examples/docker_provider/README.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+# Executing Tests
+There are some requirements in order to successfully run kitchen-terraform with Docker. Docker must be installed on the target system. This tutorial was created with Docker for Mac OS X.
```
Comment (ncs-alane) [inline examples/docker_provider/README.md]: "target system" => "system running kitchen-terraform"

Comment (ncs-alane) [inline examples/docker_provider/README.md]: The example doesn't seem to depend on the OS.

File: examples/docker_provider/main.tf [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+resource "docker_image" "chef_client" {
+  name = "${data.docker_registry_image.chef_client.name}"
+  pull_trigger = "${data.docker_registry_image.chef_client.sha256_digest}"
+  keep_locally = true
+}
+
+resource "docker_container" "chef_client" {
+  image = "${data.docker_registry_image.chef_client.name}"
+  name = "chef_client"
+  #command = [ "chef-client", "-z" ]
```
Comment (ncs-alane) [inline examples/docker_provider/main.tf]: Commented code must die :ghost:

File: examples/docker_provider/main.tf [hunk 7]
```diff
@@ -0,0 +1,23 @@
+provider "docker" {
+  host = "${var.docker_host}"
+}
+
+data "docker_registry_image" "chef_client" {
+  name = "errygg/chef-client"
```
Comment (ncs-alane) [inline examples/docker_provider/main.tf]: Can we use a more generic image?

Comment (errygg) [inline examples/docker_provider/main.tf]: Yes, I updated to use base/ubuntu:latest and it works.

File: examples/docker_provider/main.tf (hunk 6 above)
Comment (errygg) [inline examples/docker_provider/main.tf]: It is dead and gone.

File: examples/docker_provider/README.md (hunk 4 above)
Comment (errygg) [inline examples/docker_provider/README.md]: Yep, makes sense. Removed that text.

File: examples/docker_provider/README.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+[variables.tf]: variables.tf
+
+## Test Kitchen Configuration
+.kitchen.yml defines the test-kitchen configuration to include the driver, provisioner, and test groups to test.
+
+### Driver
+The driver is defined at terraform and does not have any additional configurations.
+
+### Provisioner
+The provisioner defines a variable file, [testing.tfvars], that includes the Docker configuration variables.
```
Comment (errygg) [inline examples/docker_provider/README.md]: Yep, missed that, edited the text to note using variables directly.

File: examples/docker_provider/README.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
+# Example running kitchen-terraform with the terraform docker provider
+
+This example will walk you through how to setup your environment to utilize kitchen-terraform to manage and test a local docker container.
+
+## Terraform Configuration
+
+[main.tf] defines the docker provider and associated resources.
+[main.tf]: main.tf
+
+[output.tf] defines
```
Comment (errygg) [inline examples/docker_provider/README.md]: Added descriptions

File: examples/docker_provider/main.tf (hunk 7 above)
Comment (errygg) [inline examples/docker_provider/main.tf]: Well, spoke too soon. SSH is not enabled on the base ubuntu image. Looking into how to enable that now.

Comment (errygg) [inline examples/docker_provider/main.tf]: Ok, I found a "better" one that isn't my own image: https://hub.docker.com/r/rastasheep/ubuntu-sshd/. Will post the changes soon. Doing some final testing.

File: examples/docker_provider/Gemfile [hunk 10]
```diff
@@ -0,0 +1,3 @@
+source 'http://rubygems.org/'
+
+gem 'kitchen-terraform'
```
Comment (ncs-alane) [inline examples/docker_provider/Gemfile]: This still needs to source the project code base:

```ruby
source 'https://rubygems.org/' do
  gem 'kitchen-terraform', path: '../../'
end
```
